python/ray/serve/tests/test_backend_state.py

```diff
@@ -1423,6 +1423,54 @@ def test_health_check(mock_backend_state):
     check_counts(backend_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
 
 
+def test_shutdown_through_controller(mock_backend_state):
+    """
+    Test to ensure a shutdown action goes through controller's backend 
+    state change and handled by controller's main event loop.
+    """
+    backend_state, timer, goal_manager = mock_backend_state
+
+    assert len(backend_state._replicas) == 0
+
+    b_info_1 = backend_info()
+    create_goal, updating = backend_state.deploy_backend(TEST_TAG, b_info_1)
+    assert updating
+
+    # Single replica should be created.
+    backend_state.update()
+    check_counts(backend_state, total=1, by_state=[(ReplicaState.STARTING, 1)])
+
+    # update() should not transition the state if the replica isn't ready.
+    backend_state.update()
+    check_counts(backend_state, total=1, by_state=[(ReplicaState.STARTING, 1)])
+    backend_state._replicas[TEST_TAG].get()[0]._actor.set_ready()
+    assert not goal_manager.check_complete(create_goal)
+
+    # Now the replica should be marked running.
+    backend_state.update()
+    check_counts(backend_state, total=1, by_state=[(ReplicaState.RUNNING, 1)])
+
+    assert not backend_state._replicas[TEST_TAG].get()[0]._actor.stopped
+
+    shutdown_goals = backend_state.shutdown()
+    print(shutdown_goals)
+    backend_state.update()
+
+    assert backend_state._replicas[TEST_TAG].get()[0]._actor.stopped
```

Comment:
in these unit tests we need to explicitly mark that the replicas are finished shutting down before they'll be marked stopped (mocks out the behavior of an actor exiting).

Here, we should check that force_stop was called on the replica (check the `force_stopped_counter` in the mocked actor replica) then we can mark it as exited and check that it is marked stopped and then removed.